File: modules/discovery.py
"""
Module 1: Host Discovery
Performs ARP scanning for local subnets, ICMP ping sweeps, and TCP ping probes.
"""

import asyncio
import ipaddress
import socket
import time
import subprocess
import re
from typing import List, Optional
from core.models import HostInfo
from core.logger import log_info, log_success, log_warning, log_error, log_audit
from core.storage import save_hosts

import logging
logging.getLogger("scapy.runtime").setLevel(logging.ERROR)

# Try importing Scapy
try:
    from scapy.all import ARP, Ether, srp, conf
    conf.verb = 0
    SCAPY_AVAILABLE = True
except Exception:
    SCAPY_AVAILABLE = False


def parse_targets(target_str: str) -> List[str]:
    """Parse CIDR, range, hostname, or single IP into a list of IP strings."""
    target_str = target_str.strip()
    
    # Check if target is a hostname (not IP/CIDR)
    if not any(char in target_str for char in ['/', '-']) and not re.match(r'^\d+\.\d+\.\d+\.\d+$', target_str):
        try:
            ip = socket.gethostbyname(target_str)
            return [ip]
        except socket.gaierror:
            return [target_str]

    # Handle IP Range (e.g. 192.168.1.1-192.168.1.50 or 192.168.1.1-50)
    if '-' in target_str:
        parts = target_str.split('-')
        start_ip_str = parts[0].strip()
        end_part = parts[1].strip()
        
        start_ip = ipaddress.IPv4Address(start_ip_str)
        if '.' in end_part:
            end_ip = ipaddress.IPv4Address(end_part)
        else:
            # e.g., 192.168.1.1-50
            prefix = '.'.join(start_ip_str.split('.')[:-1])
            end_ip = ipaddress.IPv4Address(f"{prefix}.{end_part}")
            
        start_int = int(start_ip)
        end_int = int(end_ip)
        return [str(ipaddress.IPv4Address(i)) for i in range(start_int, end_int + 1)]

    # Handle CIDR (e.g. 192.168.1.0/24 or single IP 192.168.1.1/32)
    try:
        network = ipaddress.ip_network(target_str, strict=False)
        if network.num_addresses == 1:
            return [str(network.network_address)]
        # Return all host IPs in the subnet
        return [str(ip) for ip in network.hosts()]
    except ValueError:
        # Fallback to single target
        return [target_str]


def arp_scan_scapy(target_cidr: str, timeout: float = 2.0) -> List[HostInfo]:
    """Perform ARP scan using Scapy for local network."""
    if not SCAPY_AVAILABLE:
        return []
    
    alive_hosts: List[HostInfo] = []
    try:
        ans, _ = srp(
            Ether(dst="ff:ff:ff:ff:ff:ff") / ARP(pdst=target_cidr),
            timeout=timeout,
            verbose=False
        )
        for _, rcv in ans:
            alive_hosts.append(
                HostInfo(
                    ip=rcv.psrc,
                    mac=rcv.hwsrc,
                    discovery_method="arp",
                    state="alive"
                )
            )
    except Exception as e:
        log_warning(f"Scapy ARP scan hatası (fallback kullanılacak): {e}")
    
    return alive_hosts


def get_arp_table() -> dict:
    """Read system ARP cache across Windows, Linux, and macOS as a fallback."""
    arp_map = {}
    try:
        if os.name == "nt":
            output = subprocess.check_output(["arp", "-a"], stderr=subprocess.DEVNULL).decode("latin-1", errors="ignore")
            for line in output.splitlines():
                match = re.search(r"(\d+\.\d+\.\d+\.\d+)\s+([0-9a-fA-F\-]{17})\s+(\w+)", line)
                if match:
                    ip, mac, _type = match.groups()
                    arp_map[ip] = mac.replace("-", ":").lower()
        else:
            # Linux / macOS
            output = subprocess.check_output(["arp", "-a"], stderr=subprocess.DEVNULL).decode("utf-8", errors="ignore")
            for line in output.splitlines():
                match = re.search(r"\(?(\d+\.\d+\.\d+\.\d+)\)?\s+(?:at\s+)?([0-9a-fA-F:]{17})", line)
                if match:
                    ip, mac = match.groups()
                    arp_map[ip] = mac.lower()
    except Exception:
        pass
    return arp_map


async def async_tcp_ping(ip: str, port: int, timeout: float = 1.0) -> Optional[float]:
    """Check if host responds to TCP connect on a specific port."""
    start_time = time.perf_counter()
    try:
        conn = asyncio.open_connection(ip, port)
        reader, writer = await asyncio.wait_for(conn, timeout=timeout)
        writer.close()
        try:
            await writer.wait_closed()
        except Exception:
            pass
        latency = (time.perf_counter() - start_time) * 1000.0
        return round(latency, 2)
    except (asyncio.TimeoutError, ConnectionRefusedError, OSError):
        # ConnectionRefusedError means the HOST is alive (sent RST packet)
        return round((time.perf_counter() - start_time) * 1000.0, 2)
    except Exception:
        return None


async def async_icmp_ping(ip: str, timeout: float = 1.5) -> Optional[float]:
    """Ping host using OS system ping asynchronously (Windows, Linux, macOS)."""
    start_time = time.perf_counter()
    try:
        if os.name == "nt":
            # Windows: ping -n 1 -w timeout_ms ip
            timeout_ms = int(timeout * 1000)
            ping_cmd = ["ping", "-n", "1", "-w", str(timeout_ms), ip]
        elif sys.platform == "darwin":
            # macOS: ping -c 1 -t timeout_sec ip
            ping_cmd = ["ping", "-c", "1", "-t", str(max(1, int(timeout))), ip]
        else:
            # Linux: ping -c 1 -W timeout_sec ip
            ping_cmd = ["ping", "-c", "1", "-W", str(max(1, int(timeout))), ip]

        proc = await asyncio.create_subprocess_exec(
            *ping_cmd,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE
        )
        stdout, _ = await proc.communicate()
        stdout_upper = stdout.upper()
        if proc.returncode == 0 and (b"TTL=" in stdout_upper or b"BYTES FROM" in stdout_upper):
            latency = (time.perf_counter() - start_time) * 1000.0
            return round(latency, 2)
    except Exception:
        pass
    return None


async def probe_single_host(
    ip: str,
    common_ports: List[int],
    timeout: float = 1.5,
    arp_map: Optional[dict] = None
) -> Optional[HostInfo]:
    """Probe a single IP with ICMP ping sweep and TCP ping probes."""
    # 1. ICMP Ping check
    icmp_latency = await async_icmp_ping(ip, timeout=timeout)
    if icmp_latency is not None:
        mac = arp_map.get(ip) if arp_map else None
        return HostInfo(
            ip=ip,
            mac=mac,
            discovery_method="icmp",
            latency_ms=icmp_latency,
            state="alive"
        )
    
    # 2. TCP Ping check across common ports
    for port in common_ports:
        latency = await async_tcp_ping(ip, port, timeout=min(timeout, 1.0))
        if latency is not None:
            mac = arp_map.get(ip) if arp_map else None
            return HostInfo(
                ip=ip,
                mac=mac,
                discovery_method=f"tcp_ping:{port}",
                latency_ms=latency,
                state="alive"
            )
            
    return None


async def discover_hosts(
    target: str,
    common_ports: Optional[List[int]] = None,
    timeout: float = 2.0,
    concurrency: int = 50,
    output_file: str = "output/hosts.json"
) -> List[HostInfo]:
    """
    Main host discovery function.
    Tries ARP scan for local CIDRs, otherwise runs async ICMP/TCP ping sweeps.
    """
    if common_ports is None:
        common_ports = [80, 443, 22, 445, 8080, 3389, 21, 23, 25, 53]
        
    log_info(f"Host Discovery başlatılıyor: Hedef='{target}'")
    log_audit("HOST_DISCOVERY_START", target, f"timeout={timeout}, concurrency={concurrency}")
    
    ip_list = parse_targets(target)
    log_info(f"Taranacak toplam hedef IP adresi sayısı: {len(ip_list)}")

    discovered_hosts: List[HostInfo] = []
    seen_ips = set()

    # 1. ARP Scan if target is a CIDR/local network
    if "/" in target and SCAPY_AVAILABLE:
        try:
            log_info("Yerel ağ için ARP taraması deneniyor...")
            arp_results = arp_scan_scapy(target, timeout=timeout)
            for h in arp_results:
                if h.ip not in seen_ips:
                    seen_ips.add(h.ip)
                    discovered_hosts.append(h)
                    log_success(f"Host bulundu (ARP): {h.ip} [{h.mac or 'No MAC'}]")
        except Exception as e:
            log_warning(f"ARP taraması atlandı: {e}")

    # 2. Fallback / Remote Ping Sweeps for remaining IPs
    remaining_ips = [ip for ip in ip_list if ip not in seen_ips]
    if remaining_ips:
        arp_cache = get_arp_table()
        semaphore = asyncio.Semaphore(concurrency)

        async def worker(ip: str):
            async with semaphore:
                res = await probe_single_host(ip, common_ports, timeout=timeout, arp_map=arp_cache)
                if res and res.ip not in seen_ips:
                    seen_ips.add(res.ip)
                    discovered_hosts.append(res)
                    log_success(f"Host bulundu ({res.discovery_method}): {res.ip} (Latency: {res.latency_ms}ms)")

        tasks = [worker(ip) for ip in remaining_ips]
        await asyncio.gather(*tasks)

    # Sort hosts by IP address
    try:
        discovered_hosts.sort(key=lambda h: ipaddress.IPv4Address(h.ip))
    except Exception:
        pass

    save_hosts(discovered_hosts, output_file)
    log_info(f"Host Discovery tamamlandı. Toplam {len(discovered_hosts)} canlı host kaydedildi: {output_file}")
    log_audit("HOST_DISCOVERY_COMPLETE", target, f"found_hosts={len(discovered_hosts)}")
    
    return discovered_hosts

File: modules/discovery.go
package modules

import (
	"bytes"
	"fmt"
	"net"
	"os/exec"
	"regexp"
	"runtime"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/specter-recon/recon-tool/core"
)

// ParseTargets parses CIDR, range, single IP, or hostname into a slice of IP strings.
func ParseTargets(target string) []string {
	target = strings.TrimSpace(target)
	if target == "" {
		return nil
	}

	// Check if hostname (no CIDR, no range, not numeric IP)
	if !strings.Contains(target, "/") && !strings.Contains(target, "-") {
		if net.ParseIP(target) == nil {
			ips, err := net.LookupIP(target)
			if err == nil && len(ips) > 0 {
				var res []string
				for _, ip := range ips {
					if ipv4 := ip.To4(); ipv4 != nil {
						res = append(res, ipv4.String())
					}
				}
				if len(res) > 0 {
					return res
				}
			}
			return []string{target}
		}
		return []string{target}
	}

	// Range (e.g. 192.168.1.1-192.168.1.50 or 192.168.1.1-50)
	if strings.Contains(target, "-") {
		parts := strings.Split(target, "-")
		if len(parts) == 2 {
			startIPStr := strings.TrimSpace(parts[0])
			endPart := strings.TrimSpace(parts[1])

			startIP := net.ParseIP(startIPStr).To4()
			if startIP != nil {
				var endIP net.IP
				if strings.Contains(endPart, ".") {
					endIP = net.ParseIP(endPart).To4()
				} else {
					// e.g. 1-50
					ipParts := strings.Split(startIPStr, ".")
					prefix := strings.Join(ipParts[:3], ".")
					endIP = net.ParseIP(fmt.Sprintf("%s.%s", prefix, endPart)).To4()
				}

				if endIP != nil {
					startInt := ipToUint32(startIP)
					endInt := ipToUint32(endIP)
					if startInt <= endInt {
						var res []string
						for i := startInt; i <= endInt; i++ {
							res = append(res, uint32ToIP(i).String())
						}
						return res
					}
				}
			}
		}
	}

	// CIDR (e.g. 192.168.1.0/24)
	ip, ipnet, err := net.ParseCIDR(target)
	if err == nil {
		var res []string
		for ip := ip.Mask(ipnet.Mask); ipnet.Contains(ip); incIP(ip) {
			res = append(res, ip.String())
		}
		// Filter network and broadcast address if network has >2 addresses
		if len(res) > 2 {
			return res[1 : len(res)-1]
		}
		return res
	}

	return []string{target}
}

func ipToUint32(ip net.IP) uint32 {
	ip = ip.To4()
	return uint32(ip[0])<<24 | uint32(ip[1])<<16 | uint32(ip[2])<<8 | uint32(ip[3])
}

func uint32ToIP(n uint32) net.IP {
	return net.IPv4(byte(n>>24), byte(n>>16), byte(n>>8), byte(n))
}

func incIP(ip net.IP) {
	for j := len(ip) - 1; j >= 0; j-- {
		ip[j]++
		if ip[j] > 0 {
			break
		}
	}
}

// GetSystemARPTable reads the OS ARP table.
func GetSystemARPTable() map[string]string {
	arpMap := make(map[string]string)
	cmd := exec.Command("arp", "-a")
	out, err := cmd.Output()
	if err != nil {
		return arpMap
	}

	lines := strings.Split(string(out), "\n")
	reWindows := regexp.MustCompile(`(\d+\.\d+\.\d+\.\d+)\s+([0-9a-fA-F\-]{17})`)
	reUnix := regexp.MustCompile(`\(?(\d+\.\d+\.\d+\.\d+)\)?\s+(?:at\s+)?([0-9a-fA-F:]{17})`)

	for _, line := range lines {
		if runtime.GOOS == "windows" {
			matches := reWindows.FindStringSubmatch(line)
			if len(matches) > 2 {
				mac := strings.ToLower(strings.ReplaceAll(matches[2], "-", ":"))
				arpMap[matches[1]] = mac
			}
		} else {
			matches := reUnix.FindStringSubmatch(line)
			if len(matches) > 2 {
				arpMap[matches[1]] = strings.ToLower(matches[2])
			}
		}
	}
	return arpMap
}

// SystemICMPPing sends a system ping packet.
func SystemICMPPing(ip string, timeout time.Duration) *float64 {
	start := time.Now()
	var cmd *exec.Cmd

	if runtime.GOOS == "windows" {
		timeoutMs := int(timeout.Milliseconds())
		cmd = exec.Command("ping", "-n", "1", "-w", strconv.Itoa(timeoutMs), ip)
	} else if runtime.GOOS == "darwin" {
		timeoutSec := int(timeout.Seconds())
		if timeoutSec < 1 {
			timeoutSec = 1
		}
		cmd = exec.Command("ping", "-c", "1", "-t", strconv.Itoa(timeoutSec), ip)
	} else {
		timeoutSec := int(timeout.Seconds())
		if timeoutSec < 1 {
			timeoutSec = 1
		}
		cmd = exec.Command("ping", "-c", "1", "-W", strconv.Itoa(timeoutSec), ip)
	}

	out, err := cmd.CombinedOutput()
	if err == nil {
		outUpper := bytes.ToUpper(out)
		if bytes.Contains(outUpper, []byte("TTL=")) || bytes.Contains(outUpper, []byte("BYTES FROM")) {
			lat := float64(time.Since(start).Nanoseconds()) / 1e6
			return &lat
		}
	}
	return nil
}

// AsyncTCPPing tests if a host is alive by connecting to a common port.
// Returns latency and whether the port was actually open (true) or actively refused via RST (false).
func AsyncTCPPing(ip string, port int, timeout time.Duration) (*float64, bool) {
	addr := net.JoinHostPort(ip, strconv.Itoa(port))
	start := time.Now()
	conn, err := net.DialTimeout("tcp", addr, timeout)
	if err != nil {
		// If connection was refused actively (RST), host is definitely alive!
		if strings.Contains(err.Error(), "refused") {
			lat := float64(time.Since(start).Nanoseconds()) / 1e6
			return &lat, false
		}
		return nil, false
	}
	_ = conn.Close()
	lat := float64(time.Since(start).Nanoseconds()) / 1e6
	return &lat, true
}

// ProbeSingleHost probes a host via ICMP and common TCP ports.
func ProbeSingleHost(ip string, commonPorts []int, timeout time.Duration, arpMap map[string]string) *core.HostInfo {
	// 1. ICMP Ping check
	lat := SystemICMPPing(ip, timeout)
	if lat != nil {
		mac := arpMap[ip]
		host := core.NewHostInfo(ip, "icmp")
		host.MAC = mac
		host.LatencyMs = lat
		return &host
	}

	// 2. TCP Ping check across common ports
	for _, port := range commonPorts {
		tcpLat, isOpen := AsyncTCPPing(ip, port, 1000*time.Millisecond)
		if tcpLat != nil {
			mac := arpMap[ip]
			method := fmt.Sprintf("tcp_open:%d", port)
			if !isOpen {
				method = fmt.Sprintf("tcp_rst:%d", port)
			}
			host := core.NewHostInfo(ip, method)
			host.MAC = mac
			host.LatencyMs = tcpLat
			return &host
		}
	}

	return nil
}

// DiscoverHosts scans and identifies all responsive hosts in the target scope.
func DiscoverHosts(target string, commonPorts []int, timeout time.Duration, concurrency int, outputFile string) ([]core.HostInfo, error) {
	core.LogInfo("Host Discovery başlatılıyor: Hedef='%s'", target)
	core.LogAudit("HOST_DISCOVERY_START", target, fmt.Sprintf("timeout=%v, concurrency=%d", timeout, concurrency), "SUCCESS")

	if len(commonPorts) == 0 {
		commonPorts = []int{80, 443, 22, 445, 8080, 3389, 21, 23, 25, 53}
	}
	if concurrency <= 0 {
		concurrency = 50
	}
	if timeout <= 0 {
		timeout = 2 * time.Second
	}

	ipList := ParseTargets(target)
	core.LogInfo("Taranacak toplam hedef IP adresi sayısı: %d", len(ipList))

	arpMap := GetSystemARPTable()

	ipChan := make(chan string, len(ipList))
	for _, ip := range ipList {
		ipChan <- ip
	}
	close(ipChan)

	var (
		wg             sync.WaitGroup
		mu             sync.Mutex
		discoveredHost []core.HostInfo
	)

	for i := 0; i < concurrency; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for ip := range ipChan {
				h := ProbeSingleHost(ip, commonPorts, timeout, arpMap)
				if h != nil {
					mu.Lock()
					discoveredHost = append(discoveredHost, *h)
					mu.Unlock()
					latStr := "N/A"
					if h.LatencyMs != nil {
						latStr = fmt.Sprintf("%.2fms", *h.LatencyMs)
					}
					core.LogSuccess("Canlı Host Bulundu: %s ➔ %s (%s, %s)", h.IP, h.State, h.DiscoveryMethod, latStr)
				}
			}
		}()
	}

	wg.Wait()

	sort.Slice(discoveredHost, func(i, j int) bool {
		ip1 := net.ParseIP(discoveredHost[i].IP).To4()
		ip2 := net.ParseIP(discoveredHost[j].IP).To4()
		if ip1 != nil && ip2 != nil {
			return ipToUint32(ip1) < ipToUint32(ip2)
		}
		return discoveredHost[i].IP < discoveredHost[j].IP
	})

	if outputFile != "" {
		_ = core.SaveHosts(discoveredHost, outputFile)
		core.LogInfo("Host Discovery tamamlandı: %d canlı host kaydedildi (%s).", len(discoveredHost), outputFile)
	} else {
		core.LogInfo("Host Discovery tamamlandı: %d canlı host tespit edildi.", len(discoveredHost))
	}
	core.LogAudit("HOST_DISCOVERY_COMPLETE", target, fmt.Sprintf("found_hosts=%d", len(discoveredHost)), "SUCCESS")

	return discoveredHost, nil
}
